bootstrap/src/stream.ml

```diff
@@ -0,0 +1,576 @@
+open Rudiments
+
+type 'a elm =
+  | Nil
+  | Cons of 'a * 'a t
+and 'a t = 'a elm lazy_t
+
+let empty = lazy Nil
+
+let init n ~f =
+  let rec fn i n f = lazy begin
+    match i < n with
+    | false -> Nil
+    | true -> begin
+      let elm = f i in
+      let t = fn (Usize.succ i) n f in
+      Cons(elm, t)
+    end
+  end in
+  fn 0 n f
+
+let length t =
+  let rec fn t i = begin
+    match t with
+    | (lazy Nil) -> i
+    | (lazy (Cons(_, t'))) -> fn t' (Usize.succ i)
+  end in
+  fn t 0
+
+let is_empty = function
+  | lazy Nil -> true
+  | lazy (Cons(_, _)) -> false
+
+let hd = function
+  | lazy Nil -> halt "Empty stream has no head"
+  | lazy (Cons(elm, _)) -> elm
+
+let tl = function
+  | lazy Nil -> halt "Empty stream has no tail"
+  | lazy (Cons(_, t)) -> t
+
+let push' elm s =
+  Cons(elm, s)
+
+let push elm s =
+  lazy (push' elm s)
+
+let pop = function
+  | lazy Nil -> halt "Empty stream has no head"
+  | lazy (Cons(elm, t)) -> (elm, t)
+
+let rec concat t0 t1 = lazy begin
+  match t0 with
+  | (lazy Nil) -> Lazy.force t1
+  | (lazy (Cons(elm, t0'))) -> push' elm (concat t0' t1)
+end
+
+let rec take t n =
+  lazy begin
+    match t, n with
+    | _, 0 -> Nil
+    | (lazy Nil), _ -> Lazy.force t
+    | (lazy (Cons(elm, t'))), _ -> push' elm (take t' (n - 1))
+  end
+
+let drop t n =
+  let rec drop' t n = begin
+    match t, n with
+    | _, 0 -> Lazy.force t
+    | (lazy Nil), _ -> Lazy.force t
+    | (lazy (Cons(_, t'))), _ -> drop' t' (n - 1)
+  end in
+  lazy (drop' t n)
+
+let split t n =
+  take t n, drop t n
+
+let rev t =
+  let rec rev' t r = begin
+    match t, r with
+    | (lazy Nil), _ -> Lazy.force r
+    | (lazy (Cons(elm, t'))), _ -> rev' t' (push elm r)
+  end in
+  lazy (rev' t empty)
+
+let rev_take t n =
+  rev (take t n)
+
+let rev_split t n =
+  rev_take t n, drop t n
+
+let pp pp_elm ppf t =
+  let open Format in
+  fprintf ppf "@[<h>";
+  let rec fn t = begin
+    match t with
+    | lazy Nil -> fprintf ppf "Nil"
+    | lazy(Cons(elm, t')) -> begin
+      fprintf ppf "(%a@ " pp_elm elm;
+      fn t';
+      fprintf ppf ")"
+    end
+  end in
+  fn t;
+  fprintf ppf "@]"
+
+let%expect_test "empty" =
+  let open Format in
+  let ppt = (pp Usize.pp) in
+  printf "@[<h>";
+  let t = empty in
+  printf "empty = %a\n" ppt t;
+  printf "@]";
+
+  [%expect{| empty = Nil |}]
+
+let%expect_test "init" =
+  let open Format in
+  let ppt = (pp Usize.pp) in
+  printf "@[<h>";
+  let rec test_init_up_to i n = begin
+    match i <= n with
+    | false -> ()
+    | true -> begin
+      let t = init i ~f:(fun i' -> i') in
```

Comment:
`i'` would be better as `i` or `j` here.  `i'` typically indicates the value is an immediate derivative of `i`.